src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -554,94 +738,53 @@ func (s *commitLogSource) bootstrapShardBlockSnapshot(
-			expectedChecksum uint32
-		)
-		if metadataOnly {
-			id, tagsIter, _, _, err = reader.ReadMetadata()
-		} else {
-			id, tagsIter, data, expectedChecksum, err = reader.Read()
-		}
+		id, tags, data, expectedChecksum, err := reader.Read()
 		if err != nil && err != io.EOF {
-			return shardResult, err
+			return err
 		}
-
 		if err == io.EOF {
 			break
 		}
 
 		dbBlock := blocksPool.Get()
-		dbBlock.Reset(blockStart, blockSize, ts.NewSegment(data, nil, ts.FinalizeHead), nsCtx)
+		dbBlock.Reset(blockStart, blockSize,
+			ts.NewSegment(data, nil, ts.FinalizeHead), nsCtx)
 
-		if !metadataOnly {
-			// Resetting the block will trigger a checksum calculation, so use that instead
-			// of calculating it twice.
-			checksum, err := dbBlock.Checksum()
-			if err != nil {
-				return shardResult, err
-			}
-
-			if checksum != expectedChecksum {
-				return shardResult, fmt.Errorf("checksum for series: %s was %d but expected %d", id, checksum, expectedChecksum)
-			}
+		// Resetting the block will trigger a checksum calculation, so use that instead
+		// of calculating it twice.
+		checksum, err := dbBlock.Checksum()
+		if err != nil {
+			return err
 		}
-
-		var (
-			tags             ident.Tags
-			shouldDecodeTags = true
-		)
-		if allSeriesSoFar != nil {
-			if existing, ok := allSeriesSoFar.Get(id); ok {
-				// If we've already bootstrapped this series for a different block, we don't need
-				// another copy of the IDs and tags.
-				id.Finalize()
-				id = existing.ID
-				tags = existing.Tags
-				shouldDecodeTags = false
-			}
+		if checksum != expectedChecksum {
+			return fmt.Errorf("checksum for series: %s was %d but expected %d", id, checksum, expectedChecksum)
 		}
 
-		if shouldDecodeTags {
-			// Only spend cycles decoding the tags if we've never seen them before.
-			if tagsIter.Remaining() > 0 {
-				tags, err = convert.TagsFromTagsIter(id, tagsIter, idPool)
-				if err != nil {
-					return shardResult, fmt.Errorf("unable to decode tags: %v", err)
-				}
-			}
+		ref, err := accumulator.CheckoutSeries(id, tags)
+		if err != nil {
+			return err
 		}
 
-		// Always close even if we didn't use it.
-		tagsIter.Close()
-
-		// Mark the ID and Tags as no finalize to enable no-copy optimization later
-		// in the bootstrap process (when they're being loaded into the shard). Also,
-		// technically we'll be calling NoFinalize() repeatedly on the same IDs for
-		// different blocks since we're reusing them, but thats ok as it an idempotent
-		// operation and there is no concurrency here.
-		id.NoFinalize()
-		tags.NoFinalize()
-
-		if shardResult == nil {
-			// Delay initialization so we can estimate size.
-			shardResult = result.NewShardResult(reader.Entries(), s.opts.ResultOptions())
+		// Load into series.
+		if err := ref.Series.LoadBlock(dbBlock, series.WarmWrite); err != nil {
+			return err
 		}
-		shardResult.AddBlock(id, tags, dbBlock)
+
+		// Always finalize both ID and tags after loading block.
```

Comment:
Do these need to be deferred? There's a few error cases above that may cause this to be skipped

Reply:
Can't really as this in a loop.  In the case of an error I think we're ok leaking some data here. GC will eventually collect these if we don't finalize.